File: strategy/python/mystrategy.py

from abc import ABC, abstractmethod

class StrategyInterface(ABC):
    def performOperation(self, a, b) -> str:
        pass
    
class OpenPanStrategy(StrategyInterface):
    def performOperation(self, a, b) -> str:
        retval = "Open Pan Strategy " + str(a + b)
        return retval

class ClosedPanStrategy(StrategyInterface):
    def performOperation(self, a, b) -> str:
        retval = "Closed Pan Strategy " + str(a * b)
        return retval

class Strategy():
    def executeStrategy(self, a, b, strategyinterface) -> str:
        return strategyinterface.performOperation(a, b)
    
class CriticalStrategy(Strategy):
    def executeStrategy(self, a, b, strategyinterface) -> str:
        print("Critica Strategy")
        return(super().executeStrategy(a, b, strategyinterface))
        
class NonCriticalStrategy(Strategy):
    def executeStrategy(self, a, b, strategyinterface) -> str:
        print("Non-Critical Strategy")
        return(super().executeStrategy(a, b, strategyinterface))

class FoodPreparation():
    def __init__(self):
        self.strategyinterface = None
    
    def setStrategy(self, strategyinterface):
        self.strategyinterface = strategyinterface
        
    def prepareFood(self):
        if not self.strategyinterface == None:
            if self.strategyinterface.__class__ == CriticalStrategy:
                a, b = 1, 2
                print(self.strategyinterface.executeStrategy(a, b, OpenPanStrategy()))
            elif self.strategyinterface.__class__ == NonCriticalStrategy:
                a, b = 3, 4
                print(self.strategyinterface.executeStrategy(a, b, ClosedPanStrategy()))    
        else:
            print("Strategy interface not set yet")
            
def main():
    foodPreparation = FoodPreparation()
    foodPreparation.setStrategy(CriticalStrategy())
    foodPreparation.prepareFood()
    
    foodPreparation.setStrategy(NonCriticalStrategy())
    foodPreparation.prepareFood()

if __name__ == "__main__":
    main()

File: strategy/swift/mystrategy.swift
import Foundation

// 1. Strategy Interface
protocol StrategyInterface {
    func performOperation(a: Int, b: Int)
}

// 2. Concrete StrategyInterface Implementations
class OpenPan: StrategyInterface {
    func performOperation(a: Int, b: Int) {
        print("Cooking in an open pan at \(a)°C for \(b) minutes.")
    }
}

class ClosedPan: StrategyInterface {
    func performOperation(a: Int, b: Int) {
        print("Pressure cooking in a closed pan at \(a)°C for \(b) minutes.")
    }
}

// 3. Strategy Class (Context/Executor)
class Strategy {
    func executeStrategy(a: Int, b: Int, strategy: StrategyInterface) {
        strategy.performOperation(a: a, b: b)
    }
}

// 4. Extensions of Strategy
class OpenStrategy: Strategy {
    override func executeStrategy(a: Int, b: Int, strategy: StrategyInterface) {
        print("Initiating Open Air Method...")
        super.executeStrategy(a: a, b: b, strategy: strategy)
    }
}

class ClosedStrategy: Strategy {
    override func executeStrategy(a: Int, b: Int, strategy: StrategyInterface) {
        print("Initiating Sealed Method...")
        super.executeStrategy(a: a, b: b, strategy: strategy)
    }
}

// 5. Food Preparation (High-level Context)
class FoodPreparation {
    private var strategy: StrategyInterface?

    func setStrategy(strategy: StrategyInterface) {
        self.strategy = strategy
    }

    func prepareFood(a: Int, b: Int, executor: Strategy) {
        guard let strategy = strategy else {
            print("No cooking strategy set!")
            return
        }
        executor.executeStrategy(a: a, b: b, strategy: strategy)
    }
}

// 6. Main Execution
func main() {
    let kitchen = FoodPreparation()
    
    // Scenario 1: Open Pan Cooking
    let openPan = OpenPan()
    let openExec = OpenStrategy()
    
    kitchen.setStrategy(strategy: openPan)
    kitchen.prepareFood(a: 180, b: 20, executor: openExec)
    
    print("---------------------------")
    
    // Scenario 2: Closed Pan Cooking
    let closedPan = ClosedPan()
    let closedExec = ClosedStrategy()
    
    kitchen.setStrategy(strategy: closedPan)
    kitchen.prepareFood(a: 120, b: 45, executor: closedExec)
}

main()
